### PR TITLE
Task 7

### PR DESCRIPTION
**Pull request by Wgheth**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -167,6 +157,60 @@ describe("app", () => {
...
+        expect(result.body[0]).toHaveProperty("votes");
+        expect(result.body[0]).toHaveProperty("author")
+        expect(result.body[0]).toHaveProperty("created_at")
+
+        expect(result.body[0].body).toBe("I have never expected this to be stressful.")
+        expect(result.body[0].author).toBe("icellusedkars")
+           
+        });
+    });
+  });
```
Comment (NRMSMITH) [inline __tests__/app.test.js:234]: Add tests for the following scenario:

the username does not exist
the required fields (username / body) are not present
201: the body sent has extra properties, i.e.: {username: 'icellusedkars', body: 'comment', hour: '12'} (hour should be ignored!)

File: controllers/news.controller.js [hunk 2]
```diff
@@ -46,4 +46,16 @@ const {
...
   }
- module.exports = {  getObjects, getArticles, getArticlesID, getComments };
\ No newline at end of file
+
+  const postComment = (req, res, next) => {
+    
+    addComment(req)
+      .then((data) => {
+        res.status(201).send(data);
+      })
```
Comment (NRMSMITH) [inline controllers/news.controller.js:55]: what you send back should be on an appropriate key, such as 'comment'.

**Review by NRMSMITH — CHANGES_REQUESTED** [review]
Req changes: a few tests need to be added (see comments!) 😄

File: __tests__/app.test.js [hunk 3]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+   .get(`/api/articles/${article_id}/commentS`)
+    .expect(404)
+    .then(({body})=>{
+    expect(body.msg).toBe("Article not found")
+
+      });
+    
+  });
+
+  test('Returns Error with 404 if the article id is invalid', () => {
```
Comment (mezzDavies) [inline __tests__/app.test.js:173]: You need to update the test description to match the assertion 404 v 400.

File: __tests__/app.test.js [hunk 4]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+    
+    return request(app)
+   .get("/api/articles/article/commentS")
```
Comment (mezzDavies) [inline __tests__/app.test.js:176]: You could also make the test more explicit by putting something like `.get("/api/articles/invalid_article_id/comment")` . No biggie - just helpful.

File: __tests__/app.test.js [hunk 5]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+    .then(({body})=>{
+    expect(body.msg).toBe("Invalid Path")
+
+      });
+    
+  });
+
   
+
+  describe("POST /api/articles/:article_id/comments", () => {
```
Comment (mezzDavies) [inline __tests__/app.test.js:187]: We don't need the array to be returned here just the created comment object (on a key of comment or similar). Arrays are for more than one item and we are only returning one item - the comment. We can make this happen by returning just the comment object from the Model to the Controller.

File: models/news.model.js [hunk 6]
```diff
@@ -39,39 +36,50 @@ const readArticlesID = (id) =>{
...
-    
 
+}
 
+function addComment(req) {
+  const { article_id } = req.params;
+  return db
+    .query(
+      "INSERT INTO comments (body, author, article_id, votes)VALUES($1,$2,$3,$4) RETURNING *;",
+      [req.body[0]["body"], req.body[0]["username"], article_id, 0]
```
Comment (mezzDavies) [inline models/news.model.js:69]: All this (taking data from the request) should be handled in the controller.

File: __tests__/app.test.js [hunk 7]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+        expect(result.body[0]).toHaveProperty("created_at")
+
+        expect(result.body[0].body).toBe("I have never expected this to be stressful.")
+        expect(result.body[0].author).toBe("icellusedkars")
+           
+        });
+    });
+
+    test("It returns status code 201 and array with an object that have an updated body and author properties", () => {
+      const postartecle = [
```
Comment (mezzDavies) [inline __tests__/app.test.js:217]: You need to send the comment that is to be added as an object not in an array here and deal with it as per my other comments (in the controller).

File: __tests__/app.test.js [hunk 8]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+        expect(result.body[0]).toHaveProperty("author")
+        expect(result.body[0]).toHaveProperty("created_at")
+
+        expect(result.body[0].body).toBe("I have never expected this to be stressful.")
+        expect(result.body[0].author).toBe("icellusedkars")
+           
+        });
+    });
+
+    test("It returns status code 201 and array with an object that have an updated body and author properties", () => {
```
Comment (mezzDavies) [inline __tests__/app.test.js:216]: The test description needs to match the assertion made. 👍

File: __tests__/app.test.js [hunk 9]
```diff
@@ -117,17 +117,7 @@ describe("app", () => {
 
 
   describe("/api/articles/:article_id/comments", () => {
```
Comment (mezzDavies) [inline __tests__/app.test.js:119]: Think about what exactly this describe block contains  - these are all GET requests

File: models/news.model.js [hunk 10]
```diff
@@ -39,39 +36,50 @@ const readArticlesID = (id) =>{
...
+    )
+    .then((result) => {
+      
+    return result.rows;
```
Comment (mezzDavies) [inline models/news.model.js:73]: You can take the comment off rows and return it back to the controller here, as per my other comments.

File: __tests__/app.test.js [hunk 11]
```diff
@@ -167,6 +157,79 @@ describe("app", () => {
...
+      return request(app)
+        .post(`/api/articles/${article_id}/comments`)
+        .send(postartecle)
+        .expect(404)
+        .then(({body})=>{
+        expect(body.msg).toBe("User does not exist")
+
+    });
+    });
+  });
```
Comment (mezzDavies) [inline __tests__/app.test.js:233]: You still need to test for: 

- if the username and or body are not present on the request 
- if the body has extra properties - they are ignored
- a request is made with valid but non existent article id
- the body is missing the required properties

**Review by mezzDavies — CHANGES_REQUESTED** [review]
Hi Waled there is still a lot of things to think about - especially with your testing. My comments should spell the way 👍 

I have also left some suggestions for tidying up test descriptions 😄 

Please have a look and resubmit.

**Pull request closed without merging**